src/transformers/models/rag/modeling_tf_rag.py

```diff
@@ -0,0 +1,1545 @@
+            >>> # initialize with RagRetriever to do everything in one forward call
+            >>> model = TFRagTokenForGeneration.from_pretrained("facebook/rag-token-nq", retriever=retriever, from_pt=True)
+
+            >>> input_dict = tokenizer.prepare_seq2seq_batch("How many people live in Paris?", "In Paris, there are 10 million people.", return_tensors="tf")
+            >>> outputs = model(input_dict, output_retrieved=True)
+
+            >>> # or use retriever separately
+            >>> # 1. Encode
+            >>> input_ids = input_dict["input_ids"]
+            >>> question_hidden_states = model.question_encoder(input_ids)[0]
+            >>> # 2. Retrieve
+            >>> docs_dict = retriever(input_ids.numpy(), question_hidden_states.numpy(), return_tensors="tf")
+            >>> doc_scores = tf.squeeze(tf.matmul(tf.expand_dims(question_hidden_states, axis=[1]), docs_dict["retrieved_doc_embeds"], transpose_b=True), axis=[1])
+            >>> # 3. Forward to generator
+            >>> outputs = model(inputs=None, context_input_ids=docs_dict["context_input_ids"], context_attention_mask=docs_dict["context_attention_mask"], doc_scores=doc_scores, decoder_input_ids=input_dict["labels"])
+
+            >>> # or directly generate
+            >>> generated = model.generate(context_input_ids=docs_dict["context_input_ids"], context_attention_mask=docs_dict["context_attention_mask"], doc_scores=doc_scores)
+            >>> generated_string = tokenizer.batch_decode(generated, skip_special_tokens=True)
+        """
+
+        assert (
+            "decoder_cached_states" not in kwargs
+        ), "Please use past_key_values to cache intermediate outputs"  # from modeling_tf_bart.py
+
+        inputs = input_processing(
+            func=self.call,
+            config=self.config,
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            decoder_attention_mask=decoder_attention_mask,
+            encoder_outputs=encoder_outputs,
+            past_key_values=past_key_values,
+            doc_scores=doc_scores,
+            context_input_ids=context_input_ids,
+            context_attention_mask=context_attention_mask,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            output_retrieved=output_retrieved,
+            n_docs=n_docs,
+            do_marginalize=do_marginalize,
+            labels=labels,
+            reduce_loss=reduce_loss,
+            training=training,
+            kwargs_call=kwargs,
+        )
+
+        inputs["do_marginalize"] = inputs["do_marginalize"] if inputs["do_marginalize"] else self.config.do_marginalize
+        inputs["reduce_loss"] = inputs["reduce_loss"] if inputs["reduce_loss"] else self.config.reduce_loss
```

Comment:
`do_marginalize` and `reduce_loss` should be handled in the processing function not here.

Reply:
@jplu Sorry for late reply (just finished `TFRagSequenceForGeneration`). Do I understand correctly that you meant I should "modify" `input_processing` in `modeling_tf_utils` to handle `n_docs`, `do_marginalize` and `reduce_loss`  ??

Reply:
This model will never be able to be executed in graph mode, so you can forget about this. Let's say it will be the exception.